src/coreComponents/constitutive/fluid/BlackOilFluid.cpp

```diff
@@ -31,76 +26,443 @@ using namespace dataRepository;
+        real64 const viscSlope_dn = ( viscInterp_dn[i] - viscInterp_dn[i - 1] ) / ( ppTarget[i] - ppTarget[i - 1] );
+        real64 const viscSlope = ( 1 / dRs_up * viscSlope_up + 1 / dRs_dn * viscSlope_dn ) / ( 1 / dRs_up + 1 / dRs_dn );
+        m_PVTO.m_undersaturatedViscosity[iCurrent].emplace_back( m_PVTO.m_undersaturatedViscosity[iCurrent][i - 1] + viscSlope * ( ppTarget[i] - ppTarget[i - 1] ) );
+      }
+
+    }
+    upperBranchIndex = iCurrent;
+  }
 }
 
+void BlackOilFluid::extendUndersaturatedProperties()
+{
+  // extrapolate m_undersaturated properties up to max pressure
+  for( localIndex i = 0; i < m_PVTO.m_nSaturatedPoints; ++i )
+  {
+    real64 const dPext = m_PVTO.m_maxRelativePressure - m_PVTO.m_undersaturatedPressure[i].back();
+
+    array1d< real64 > const & Pusat = m_PVTO.m_undersaturatedPressure[i];
+    array1d< real64 > const & Bousat = m_PVTO.m_undersaturatedBo[i];
+    array1d< real64 > const & viscusat = m_PVTO.m_undersaturatedViscosity[i];
+
+    if( std::fabs( dPext ) > 0 )
+    {
+      localIndex const branchSize = m_PVTO.m_undersaturatedPressure[i].size();
+      real64 const Bo = linearExtrapolation( Pusat[branchSize - 2],
+                                             Bousat[branchSize - 2],
+                                             Pusat[branchSize - 1],
+                                             Bousat[branchSize - 1],
+                                             m_PVTO.m_maxRelativePressure );
+      real64 const visc = linearExtrapolation( Pusat[branchSize - 2],
+                                               viscusat[branchSize - 2],
+                                               Pusat[branchSize - 1],
+                                               viscusat[branchSize - 1],
+                                               m_PVTO.m_maxRelativePressure );
+      m_PVTO.m_undersaturatedBo[i].emplace_back( Bo );
+      m_PVTO.m_undersaturatedViscosity[i].emplace_back( visc );
+      m_PVTO.m_undersaturatedPressure[i].emplace_back( m_PVTO.m_maxRelativePressure );
+    }
+  }
 }
 
-void BlackOilFluid::postProcessInput()
+void BlackOilFluid::refineTableAndCopy( localIndex const nLevels )
 {
-  // TODO maybe use different names?
-  m_componentNames = m_phaseNames;
+  std::vector< real64 > refineP; // TODO:change to array1d
+  for( localIndex i = 0; i < m_PVTO.m_nSaturatedPoints; ++i )
+  {
+    for( localIndex j = 0; j < m_PVTO.m_undersaturatedPressure[i].size(); ++j )
+    {
+      refineP.emplace_back( m_PVTO.m_undersaturatedPressure[i][j] );
+    }
+  }
+  std::sort( refineP.begin(), refineP.end() );
+  refineP.erase( std::unique( refineP.begin(), refineP.end() ), refineP.end() );
+
+  // linspace
+  real64 const maxVal = *std::max_element( refineP.begin(), refineP.end() );
+  array1d< real64 > refinePP( refineP.size() );
+  for( localIndex i = 0; i < localIndex( refineP.size() ); ++i )
+  {
+    refinePP[i] = refineP[i];
+  }
```

Comment:
Is this part even needed? It looks like `refinePP` is being overwritten entirely by the code that follows.